### PR TITLE
gh-108520: Fix bad fork detection in nested multiprocessing use case

### PR DESCRIPTION
**Pull request by albanD**
gh-107275 introduced a regression where a SemLock would fail being passed along nested child processes, as the `is_fork_ctx` attribute would be left missing after the first deserialization.

* Issue: gh-108520

### CODE REVIEW

**Comment by albanD** [comment]
I'm not sure how to read the test failure? Are they actually relevant?

**Comment by albanD** [comment]
Rebased to check fi the windows test failure was just flaky.

File: Lib/multiprocessing/synchronize.py
Comment (pitrou) [inline Lib/multiprocessing/synchronize.py]: Why not instead set `is_fork_ctx` in `__setstate__` below?

File: Lib/test/_test_multiprocessing.py
Comment (pitrou) [inline Lib/test/_test_multiprocessing.py]: Note that you can make those classmethods just above the associated test method.

File: Lib/test/_test_multiprocessing.py
Comment (pitrou) [inline Lib/test/_test_multiprocessing.py:5457]: Can you please add an explanatory comment pointing to the GH issue?

File: Lib/multiprocessing/synchronize.py
Comment (albanD) [reply Lib/multiprocessing/synchronize.py]: I'm happy to change this. My question would be: is it guaranteed that either `__init__()` or `__setstate__()` will be called for every object? Or are there other ways the object might be created that would end up not calling either of these methods?

If there is no other way, I can update `__setstate__` but if there are other ways the objects could be created, this pattern will cover them all.

Let me know which one is best and I can update the PR accordingly!

Comment (pitrou) [reply Lib/multiprocessing/synchronize.py]: Perhaps grepping through uses of `SemLock` in the Python stdlib would give an insight? This is not a public API, so external use should not be relevant.

Comment (albanD) [reply Lib/multiprocessing/synchronize.py]: Ok, so for my general questoin: for any python object, is it possible to create it without calling either `__init__()` or `__setstate__()`?
The answer is definitely yes as the user can do whatever they want with calling `__new__` directly.

In this particular, case, since SemLock is the base class for many public API (and thus user might create these objects arbitrarily) it is best to make `__getstate__` resilient to bad class creation.

Comment (pitrou) [reply Lib/multiprocessing/synchronize.py]: Well, perhaps it's possible to create an object in that way, but that is not a supported API _and_ it probably gives you a broken object already.

Comment (albanD) [inline Lib/multiprocessing/synchronize.py]: I just realized that this is the base class for many public objects like Lock.
This means that the new attribute added in the previous PR adds new (non documented) public API. Should I fix that by renaming the field `_is_fork_ctx` to make it private (in this PR or a follow up, as you prefer)? Or should I add documentation for it somewhere if we want to make it fully public?

Comment (pitrou) [reply Lib/multiprocessing/synchronize.py]: Good catch, then let's please make it `_is_fork_ctx`.

**Comment by albanD** [comment]
All the updates are done.
Let me know if the a blurb is needed?
Also I guess commits are squashed before merging? let me know if you want me to cleanup the commits in the branch.

**Comment by pitrou** [comment]
Yes, a blurb is needed especially as it fixes a regression.
Yes, commits will be squashed automatically when merging :-)

**Review by pitrou — APPROVED** [review]
+1, thanks for the quick fix @albanD

File: Lib/multiprocessing/synchronize.py
Comment (pitrou) [inline Lib/multiprocessing/synchronize.py]:
```suggestion
        # Ensure that deserialized SemLock can be serialized again (gh-108568).
```

File: Lib/multiprocessing/synchronize.py
Comment (pitrou) [inline Lib/multiprocessing/synchronize.py]:
```suggestion
        # Ensure that deserialized SemLock can be serialized again (gh-108520).
```

**Comment by pitrou** [comment]
Ok, it took me a couple attempts to get the comments right :-) Will merge if CI is green.

**Comment by vstinner** [comment]
The test fails randomly. Can you please have a look? I proposed a fix.

* Issue gh-109706
* PR gh-109707

**Pull request merged**
